clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -756,19 +737,33 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
                 }
 
                 List<ConsumerRecord<K, V>> parsed = new ArrayList<>();
+                boolean skippedRecords = false;
                 for (LogEntry logEntry : partition.records.deepEntries()) {
                     // Skip the messages earlier than current position.
                     if (logEntry.offset() >= position) {
                         parsed.add(parseRecord(tp, logEntry));
                         bytes += logEntry.sizeInBytes();
-                    }
+                    } else
+                        skippedRecords = true;
                 }
 
                 recordsCount = parsed.size();
 
                 if (!parsed.isEmpty()) {
                     log.trace("Adding fetched record for partition {} with offset {} to buffered record list", tp, position);
                     parsedRecords = new PartitionRecords<>(fetchOffset, tp, parsed);
+                } else if (!skippedRecords && (partition.records.sizeInBytes() > 0)) {
+                    // We did not read a single message from a non-empty buffer, because that message's size is
```

Comment:
Since you have edited the KIP now, can you please update the comment and error message here? I think that's the only thing left before we can merge this PR.